### ISSUE TITLE
Support reCAPTCHA component

### ISSUE THREAD

**Issue by lisotton** [ampproject/amphtml#2273]
Hi everyone, 

I don't know if someone is already working in this request, but I believe that we need to implement support for recaptcha in AMP.

Thanks.

**Comment by jmadler** [comment]
Is there a specific problem that would solve for you?  I'd like to better understand the use case.

**Comment by abhishekgoel137** [comment]
Recaptcha is required for every form.

**Comment by rudygalfi** [comment]
@jasti Do you know anyone on the captcha team we could talk to? This is going to be needed for some landing page and e-commerce scenarios.

**Comment by jasti** [comment]
We've reached out to the reCaptcha team and will update progress here.

**Comment by weeblr** [comment]
HI

Hope this can be done. Now that we have forms conversion (from html to amp) working, people need to make special version of their forms to remove the captcha field, big annoyance, and of course, there is no captcha to spam rate can be high.

Rgds

**Comment by jasti** [comment]
@weeblr Yup, we are actively looking into this.

**Comment by nickmoline** [comment]
Can we have an implementation more generic than this, something like `<amp-captcha>` that could be extended to multiple CAPTCHA providers including self provided ones.

**Comment by alanorozco** [comment]
reCAPTCHA component is currently blocked due to issues with rendering inside a 3p iframe. Will update this once we reach a decision on technical issues re: resizing and signals.

@nickmoline Sounds good, butCould c you provide an example for how a different CAPTCHA provider would work? The challenge with this is designing an API that is general enough but not badly overloaded. Would like to have more info :)

**Comment by nickmoline** [comment]
a Lot of simpler captchas work by simply generating a one time token which is used both in an image url, audio url and a hidden input field.  We could use a JSON endpoint that would return the token something like this:

```html
<form method="post" action-xhr="https://example.com/contact" target="_top">
<!-- normal form fields -->
<amp-captcha src="https://data.com/captcha.json">
   <template type="amp-mustache">
      <input type="hidden" name="captcha-token" value="{{token}}" />
      <amp-img src="{{imageUrl}}" width="100" height="50"></amp-img>
      <input type="text" name="captcha" value="" placeholder="Enter letters from the image above or from the audio file below" />
      <a href="{{audioUrl}}">Audio version of CAPTCHA</a>
   </template>
</amp-captcha>
</form>
```

the json endpoint can simply return a standard hash of values that can be used in the mustache template

```json
{
  "token": "jkljqwe12351234asfd978!",
  "imageUrl": "https://example.com/captcha/image/jkljqwe12351234asfd978!.png",
  "audioUrl": "https://example.com/captcha/audio/jkljqwe12351234asfd978!.mp3"
}
```

**Comment by NoahWilcox** [comment]
Thought I'd throw a vote in for this.

**Comment by alanorozco** [comment]
@NoahWilcox 

If you're talking about the reCAPTCHA component, no more voting should be needed :)

We're actively working on this and you can expect this issue to be updated soon.

**Comment by machadoug** [comment]
@alanorozco 
What about support for other captcha providers such as phpcaptcha.org?

The amp-captcha tag suggested by @nickmoline would be nice, however it would be just missing a way to refresh the captcha image.

**Comment by nickmoline** [comment]
Just thought I'd drop a little note here for anyone else who likes my proposed situation for a generic captcha other than recaptcha.  We ended up experimenting to see if this would work just using the existing `<amp-list>` inside of an `amp-form`  as opposed to a brand new component, and sure enough it works just fine.

```html
<form method="post" action-xhr="https://example.com/contact" target="_top">
<!-- normal form fields -->
<amp-list src="https://example.com/captcha/captcha.json" width="auto" height="60" template="amp-template-captcha" layout="fixed-height">
   <template type="amp-mustache" id="amp-template-captcha">
      <input type="hidden" name="captcha-token" value="{{token}}" />
      <amp-img src="https://example.com/captcha/image/{{token}}" width="100" height="50"></amp-img>
      <input type="text" name="captcha" value="" placeholder="Enter letters from the image above or from the audio file below" />
      <a href="https://example.com/captcha/audio/{{token}}">Audio version of CAPTCHA</a>
   </template>
</amp-list>
</form>
```

```json
{
  "token": "jkljqwe12351234asfd978",
}
```

Note this would have worked fine using my discrete fields from my example above, but in this particular example, since the filename was based on the token, may as well save the extra bytes in the .json transfer :)

---
@machadoug I would assume in order to give an option to reforce a refresh of the captcha image you could either do something with `<amp-live-list>` or `<amp-bind>` to allow for changing the element after the page loads. and perhaps an eventual `<amp-captcha>` component could support this built in, but it might be possible with `<amp-bind>` now, would have to experiment.

**Comment by jasti** [comment]
@nickmoline Glad to see this is working without a generic component. Keep us posted.

**Comment by ampprojectbot** [comment]
This issue hasn't been  updated in awhile. @alanorozco Do you have any updates?

**Comment by twd3** [comment]
Can we please get this going? I am using iframe so I can have captcha and it sucks.

**Comment by lswang1618** [comment]
Work for this is in progress! Current time frame is to have support in April - will update this issue if that changes.

**Comment by shishirm** [comment]
Hello. Just checking in to see the proposed timeline. Is this still scheduled for release this month? Is there a beta version that we can try out using dev flags or will this be released live once it is ready.

**Comment by aghassemi** [comment]
@ericlindley-g I don't believe the team that was working on this is still actively doing so, do you know?

**Comment by ericlindley-g** [comment]
I'm not sure about the status, but I think @rudygalfi might be.

**Comment by rudygalfi** [comment]
We are currently looking to support the [latest version of recaptcha (v3)](https://developers.google.com/recaptcha/docs/v3) once it's released.

**Comment by rudygalfi** [comment]
Reassigning to @lannka. It's looking like we'll get started in mid-August.

**Comment by 88kbbq** [comment]
Why not just whitelist www.google.com/recaptcha/api.js and then create <amp-recaptcha> for the v3 script. This seems like you could do this in a couple hours, why wait until August?

**Comment by lannka** [comment]
@88kbbq we can't just let any non-AMP script run in the AMP window. @torch2424 will take on this work in next sprint. Sorry for the wait.

**Comment by 88kbbq** [comment]
Well, it’s a google script so...

**Comment by lannka** [comment]
@88kbbq AMP is an open source project, and any scripts running inside AMP window has to be open sourced and visible in AMP repo.

**Comment by IamAlta** [comment]
@aghassemi @rudygalfi 
Any chance of expediting this? We're getting numerous spam, some that is obnoxious.

**Comment by lannka** [comment]
@IamAlta sorry for the wait. It's already planned for this sprint. 
To set up expectation, we'll be only supporting the latest version of ReCaptcha, which is [V3](https://developers.google.com/recaptcha/docs/v3)

**Comment by torch2424** [comment]
Hello! I'll be the one owning this feature for now, and we discussed the feature at our [Design Review](https://github.com/ampproject/amphtml/issues/16693). And I also wanted to share our Design Doc. Which will incorporate two parts: *AMP Async Inputs* and `<amp-recaptcha-input>`.

Where we currently have a rough [initial implementation](https://jsfiddle.net/torch2424/26wahxgy/), though this does not reflect end user use case.

# AMP Async Input

## Project Goal(s)
To allow for asynchronous tasks to be run before submitting an AMP form.

## Proposed Solution
The proposed solution to create a new derived class from `BaseElement`, named `AsyncInputElement`. `AsyncInputElement` should be considered as an AMP element, however with an overridable `getValue()` function that returns a promise. This will allow AMP form to query for `AsyncInputElemens` elements using an `i-amphtml-asyncinput` class query selector, and then use their overridable functions to wait for their value to be resolved and submitted alongside the form.

## Proposed / Example Class Outline
### `async-input.js`

```
// Imports

export class AsyncInputElement extends BaseElement {
  
  /** @override */
  buildCallback() {
    this.classList.add(‘i-amphtml-asyncinput’);
  }

  /**
  * Function to return the input value
  * @returns Promise<*>
  */
  getValue() {}

}
```

### Example Async Input Component

```
// Imports
export class AmpExampleAsyncInput extends AsyncInputElement {
   
  ...
  
  /** @override */
  buildCallback() {
    super.buildCallback();
    
    ...
  }

  /** @override */
  getValue() {
    ...
  }

  ...
}
```

## High-Level Implementation Walkthrough

1. Create a new AMP Component.
2. AMP Component extends `AMP.AsyncInputElement` instead of `AMP.BaseElement`
3. AMP Component overrides the `AMP.AsyncInputElement` functions, and returns promises for each.
4. End User submits an AMP Form.
5. AMP form adds loading state classes to the form.
6. AMP form scans all child elements, specifically grabbing input elements.
7. AMP form also scans for child elements, with a special class, `i-amphtml-asyncinput`.
8. AMP form finds all `Async Input` elements, and calls `.getValue()` and waits for Promises to resolve.
9. AMP form appends the resolved values to the form.
10. AMP form submits.
11. AMP form transitions away from loading state classes, and does a full page reload, or success/error state depending upon usage.

## Concerns / Edge Cases

### How will this work with AMP form’s existing Async Variable Substitution?
`AsyncInput` and variable substitution can be executed parallel, and can act as separate entities. We can simply replace the variable substitution async call, with a call to both `Async input` and variable substitution wrapped in a `Promise.all()` in the form submission flow.

### How should Errors be handled with Async Inputs?
`AsyncInput` can follow the same error flow as a rejected variable substitution. Currently, on error of variable substitution, the form is simply thrown into the error state, and is not submitted.

### Does the form currently have support for AMP components as children?
Yes, see the `<amp-selector>` component inside of forms.

### Should there be an `isReady()` function for the AsyncInput Class?
`AsyncInput` should not need an `isReady()` function, as waiting for the readiness of an input can simply be part of the `getValue()` promise of the AsyncInput component itself.

## Task Breakdown
* Create a `getAsycnInputFields_()` function.
  * This will query for all async input elements that a child of the form element, using the class `i-amphtml-asyncinput`. For example: `document.querySelectorAll(‘.i-amphtml-asyncinput’)`
* Check for async input in the `submit_` function, using `getAsyncInputFields)()`.
* If we have Async Inputs:
  * Update the `submit_` function so that the form enters the `SUBMITTING` state if there are async inputs.
  * Create a promise that calls the respective `getValue()` function of the Async Inputs.
  * Continue the form `submit_()` flow.
* If we do not have Async Inputs:
  * Follow the regular form `submit_` flow.
  * Write tests for the `getAsyncInputFields()`.
  * Write tests for the new `submit_()` flow.

**Comment by torch2424** [comment]
Hello!

Currently we have no "set date" estimate, but I can let you know on what's been happening at a high level:

So the component is definitely making its way along! We've had a few PRs make it into Master, e.g #17971 . Which makes significant progress on the component 😄 However, after working with the recaptcha team, we had to find a solution for supporting both Cannonical (`https://example.com`) and Cache (`https://example-com.cdn.ampproject.org`) domains for all AMP Caches. And thankfully we did! So currently I have been working on a module to handle getting the cache subdomain (`example-com`) inside of an AMP extension, that way we can move forward on this and get the component finished up.

After this is all code complete, we are going to put it into experimental, that way pages can start to be built, and test out the component, so we can then get feedback from developers on anything we might have missed, and then we can send it out to production.

Hope that helps! Let me know if I can help with anything else! Definitely still working on this and is one of my main focuses right now on the project.

**Comment by torch2424** [comment]
You are welcome! Sorry for the slower reply this time around, had a busy weekend 😄 

> QUESTION 1:
So, would this work?
Add a random image with the recaptcha characters to the form input.
Have visitors enter the chars into a text input.
Submit the form with submit-success and submit-error.
If the recaptcha doesn't match, then send a submit-error back.

By "recaptcha characters", are you referring to the older recaptcha that had you solve a challenge by reading modified text, and then entering the text? Also, are you asking if this would work as a temporary solution? If yes to both, I am not entirely sure, as someone could just as easily send requests directly to your server, without going through your AMP page as a normal user, thus bypassing the process. Especially in the case that there is a single image.

> QUESTION 2:
Will the visitor have a second chance to enter the form with the recaptcha
or will the program have to reload the page for them?

AMP recaptcha will only support using [reCAPTCHA v3](https://developers.google.com/recaptcha/docs/v3), which has no UI or client side challenge associated with it. And will work along side AMP form, so any type of rejection or error you would normally do with an AMP form, should also work when processing the `<amp-recaptcha-input>` inside of the form. 😄 

> QUESTION 3:
Normally, the form loads a random image upon page _reload.
Will the recaptcha image reload with a new one when the visitor is sent back with the submit-error?
If not, how do you recommend that the image be reloaded, preferably without refreshing the page (for smoothness)?

See my above response, the newest version of reCAPTCHA has no image associated 😄 

Hope this helps, feel free to continue this discussion, and you are welcome. 👍

**Comment by IamAlta** [comment]
As I understand it, AMP doesn't have reCaptcha ready.

ANSWER TO QUESTION 1.
By "recaptcha characters" ... I have images of the characters to be entered into a text input box.

1. One image is shown and the visitor enters the chars into a text input and they eventually submit the form.

2. The action-xhr URL on the form decodes which characters are correct and proceeds if they are ok.
I was trying to send back a message to display depending on whether the captcha chars matched. 
But amp-mustache doesn't work. (action-xhr, amp-mustache per https://ampbyexample.com/components/amp-form/, "Form Submission with Page Update" but the returned json is not displayed in the mustaches. (I have a request for help about that).

ANSWER TO QUESTIONS 2 and 3.
I think the form would have to _reload with a new image unless the amp-img url can be used in a mustache, then json can display a new amp-img (and the json/mustaches work)?

I am uncertain if that would work, but anyway, the visitor would be instructed to refresh the page.
Their inputs other than captcha can be saved in a session variable to load when the page refreshes.

Does this make sense?

**Comment by IamAlta** [comment]
Hi Aaron @torch2424,

As I understand it, AMP doesn't have reCaptcha ready.

ANSWER TO QUESTION 1.
By "recaptcha characters" ... I have images of the characters to be entered into a text input box.

    One image is shown and the visitor enters the chars into a text input and they eventually submit the form.

    The action-xhr URL on the form decodes which characters are correct and proceeds if they are ok.
    I was trying to send back a message to display depending on whether the captcha chars matched.
    But amp-mustache doesn't work. (action-xhr, amp-mustache per https://ampbyexample.com/components/amp-form/, "Form Submission with Page Update" but the returned json is not displayed in the mustaches. (I have a request for help about that).

ANSWER TO QUESTIONS 2 and 3.
I think the form would have to _reload with a new image unless the amp-img url can be used in a mustache, then json can display a new amp-img (and the json/mustaches work)?

I am uncertain if that would work, but anyway, the visitor would be instructed to refresh the page.
Their inputs other than captcha can be saved in a session variable to load when the page refreshes.

Does this make sense?

**Comment by IamAlta** [comment]
Hi Aaron @torch2424,

As I understand it, AMP doesn't have reCaptcha ready.

ANSWER TO QUESTION 1.
By "recaptcha characters" ... I have images of the characters to be entered into a text input box.

One image is shown and the visitor enters the chars into a text input and they eventually submit the form.

The action xhr URL on the form decodes which characters are correct and proceeds if they are ok.
I was trying to send back a message to display depending on whether the captcha chars matched.
But amp-mustache doesn't work. (action-xhr, amp-mustache per https://ampbyexample.com/components/amp-form/, "Form Submission with Page Update" but the returned json is not displayed in the mustaches. (I have a request for help about that).

ANSWER TO QUESTIONS 2 and 3.
I think the form would have to _reload with a new image unless the amp-img url can be used in a mustache, then json can display a new amp-img (and the json/mustaches work)?

I am uncertain if that would work, but anyway, the visitor would be instructed to refresh the page.
Their inputs other than captcha can be saved in a session variable to load when the page refreshes.

Does this make sense?

**Comment by torch2424** [comment]
@IamAlta 

Hello!

> As I understand it, AMP doesn't have reCaptcha ready.

Yes you are correct, Currently, AMP does not have reCAPTCHA support, though we are working on it 😄 

Now, regarding your Answers to my answers, I think what you are trying to do is implement your own [reCAPTCHA v2](https://developers.google.com/recaptcha/docs/display) as a temporary solution until `<amp-recaptcha-input>` is ready ? 

My apologies if I am having trouble understanding. 😃

**Comment by IamAlta** [comment]
@torch2424,
Almost. 
Not reCAPTCHA at all. 
Our own simple one to stop the spammers until AMP recaptcha is completed.

I think what I outlined will work but the json/mustache isn't behaving for a return message. I'll wait for that response on another issue.

You've been great - stay that way!

**Comment by torch2424** [comment]
Glad I can help! We are working to get `amp-recaptcha-input` out as soon as we can. Will definitely let you know once it hits experimental 😄

**Comment by torch2424** [comment]
Hello!

So a quick update on anyone following this, here is a quick status update:

Let me start by saying, **In my opinion, this still needs a couple weeks before experimenting with this in your development environment, due to our release cycles.**

So as of now, as far as I know, this is now available in PROD under an experiment. Meaning you can now use this for development, by doing `AMP.toggleExperiment('amp-recaptcha-input')`, but will not work for your end users. And, here are some **Gotchas**:

1. Only the [reference documentation](https://github.com/ampproject/amphtml/blob/master/extensions/amp-recaptcha-input/amp-recaptcha-input.md) is done. We are currently working on our new AMP homepage (amp.dev), and thus it is kind of / sort of blocking us on releasing new AMP By Example, examples. Currently, we have to write two versions for each, which I could do, but there are other things that need to also be done for recaptcha, so I'm hoping that gets released to lower the amount of overall work. If you are REALLY excited to try this out (as am I 😄 ), I opened a PR for the local dev with a [full end-to-end example](https://github.com/ampproject/amphtml/pull/20176). But I closed per some offline discussions, and other amp.dev stuff. This is not an official example, and is more for someone interested.

2. There are some Bugs that were fixed after this hit PROD under the experiment. And these bugs won't be in PROD under the experiment for a little while. [This one about NonXHR GET requests](https://github.com/ampproject/amphtml/pull/20362). [This one after our security review](https://github.com/ampproject/amphtml/pull/20358).

3. [Validator changes](https://github.com/ampproject/amphtml/pull/20143) were only recently added, thus they may/may not be in PROD under the experiment.

4. We are still working on an integration test.

I've had some people reach out saying that they got this working on their dev environment for their sites, and tested it working. There should be a checkbox to enable AMP support for your reCAPTCHA sites, though I also heard it was not visible yet, so your mileage may vary with that.

And most of all, anyone who plans to try this it, **please please please leave some feedback for this experiment on this issue**. The more feedback we get on the extension, the sooner we can get it out to your end users as we will feel more confident on making the extension publicly available.

Thank you everyone for your patience, and I will give another status update once we figure out the example and things, and we are certain everything is available in PROD under the experiment.

Thank you!

**Comment by torch2424** [comment]
So, another update, found out there is an SSL Error with our recaptcha iframe. Will write up a fix for that, and then let everyone know when this is 100% ready for feedback and testing 😄 

Thank you!

**Comment by torch2424** [comment]
Hello!

So the SSL Error should now all be fixed up, and all functionality that is required for `amp-recaptcha-input` should now be fixed/working in our PROD rollout of the extension! 😄 🎉 

Now all we need is AMP community feedback!

You can now use this for development, by doing `AMP.toggleExperiment('amp-recaptcha-input')`, but **`amp-recaptcha-input` will not currently work for your end users**. Meaning we'd love it if you could play around with this in your non-production environments and/or demos and things. And there should now be a checkbox on your recaptcha admin dashboard to "Allow this key to work with AMP pages". Which should enable your reCAPTCHA sitekey to be used on AMP pages using the extension.

We currently have [reference documentation](https://github.com/ampproject/amphtml/blob/master/extensions/amp-recaptcha-input/amp-recaptcha-input.md), and I made a quick example using Glitch: 

* [AMP Page Demo](https://torch2424-amp-recaptcha-example.glitch.me/) 
* [AMP Page Source](https://glitch.com/edit/#!/torch2424-amp-recaptcha-example)
* [Express Server Source, see server.js and the endpoint /api/recaptcha](https://glitch.com/edit/#!/torch2424-amp-glitch-express)

And again, anyone who plans to try this it, **please please please leave some feedback for this experiment on this issue**. The more feedback we get on the extension, the sooner we can get it out to your end users as we will feel more confident on making the extension publicly available.

Sorry the back and forth on the comments I made above about how this may be ready for development purposes 🙃. But now it definitely is! Thank you everyone for you patience! Super excited that we are close to bringing this into your AMP pages! 😄 🎉

**Comment by torch2424** [comment]
Hello Everyone!

`<amp-recaptcha-input>` is now fully available and can be deployed to AMP pages, as we launched this for AMP Conf 2019 😄 🎉 

Here are some resources to get started:

* [amp.dev reference documentation](https://amp.dev/documentation/components/amp-recaptcha-input)

* [amp.dev simple example](https://amp.dev/documentation/examples/components/amp-recaptcha-input)

* [And of course, the official announcement given by @jeffjose 😄](https://youtu.be/wPW-kXsONqA?t=272)

Thank you everyone for your patience! 👍 🙌  Looking forward to y'all trying this out! 

cc @lisotton @neluzhin @ValentinBrclz @denis90 @NoahWilcox @xileftenurb @ymkjp @SweetLittleParsley @neos1993 @talmog @ajeetku  @weeblr @nebuso  @nickmoline @machadoug @twd3 @shishirm @IamAlta @natoinet @jorenham @SeanC-Nimble @joshcp @chigia001 @AOlivar @luisfranco @BaraCoda @doppl3r @heitorfig

**Comment by McAlex220** [comment]
Hi!
Can you please help me to install <amp-recaptcha-input>?
I've put ```
<amp-recaptcha-input layout="nodisplay"
      name="recaptcha_token"
      data-sitekey="6LfcQ7IUAAAAAIv1KcgqyExGK0v8dLJtvV_Q6xD-"
      data-action="recaptcha_example">
    </amp-recaptcha-input> 
```
to index.html file. But i don't understand how to execute captcha, and how to get $_POST['g-recaptcha-response'](like in normal sites) in my form action php file to validate form...

**Comment by torch2424** [comment]
cc @zhouyx Can you help out @McAlex220 ? :smile:

**Comment by zhouyx** [comment]
Thank you @torch2424 for adding me : )

@McAlex220 I'm happy to help. Here's the docs that can be of help 
https://github.com/ampproject/amphtml/blob/master/extensions/amp-recaptcha-input/amp-recaptcha-input.md

Also the examples that @torch2424 added
* [AMP Page Demo](https://torch2424-amp-recaptcha-example.glitch.me/) 
* [AMP Page Source](https://glitch.com/edit/#!/torch2424-amp-recaptcha-example)
* [Express Server Source, see server.js and the endpoint /api/recaptcha](https://glitch.com/edit/#!/torch2424-amp-glitch-express)

I am traveling and attending conference this week. I'd be happy to help you further debug next week if you still have issue using the component. Thank you!

**Comment by torch2424** [comment]
Thank you @zhouyx ! 😄

**Comment by Frenchomatic** [comment]
It is frustrating when something like recaptcha is offered for AMP pages but nobody seems to give an example or a clear explanation of what is needed. There is not much on the web that I can find which actually describes the crucial step of what code needs to go into the server side php file. 

So in my case , in my amp page form I have this 
```
<input type="submit" value="Envoyer email">
<amp-recaptcha-input layout="nodisplay"
      name="recaptcha_token"
      data-sitekey="XXXXXXXXXXX"
      data-action="recaptcha_example">
    </amp-recaptcha-input>
```
    
I have the necessary scripts listed in the head as given on the AMP reference pages. The page validates. 

However what goes into my receiving php file. The form works AMP wise in receiving emails but I do not know what to put into that php file in order for it to support recaptcha 3.

Any help is much appreciated - this is not a support issue but one where AMP is currently allowing mass spamming as many webmasters can't implement a recaptcha. I see lots of people asking for the same as I am.

**Comment by lannka** [comment]
@Frenchomatic did you find something useful on [reCaptcha's site](https://developers.google.com/recaptcha/docs/settings)? It should not be similar (if not the same) how you integrate reCaptcha on your non-AMP pages.

**Comment by Frenchomatic** [comment]
NO - that is the point - there is no explanation of what is needed in term of the php server side script. For example, there is a secret key but what do you do with it?

**Comment by Frenchomatic** [comment]
@ValentinBrclz If you are going to act like a clown then don't get upset when you are treated like one. The documentation is poor, spread out and incomplete. There is no clear procedure to explain how to implement recaptcha version 3 on an AMP page. Perhaps you may like to fill the gap with your knowledge or proof otherwise.

**Comment by luisfranco** [comment]
> NO - that is the point - there is no explanation of what is needed in term of the php server side code. For example, there is a secret key but what do you do with it? Is it the same as what is needed in a non-AMP page, different or what? Is it something like this perhaps - I am frankly lost because all examples are incomplete. They all talk about the html file but not what needs to happen on the backend.
> 
> ```
> if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_POST['recaptcha_response'])) {
> 
>     // Build POST request:
>     $recaptcha_url = 'https://www.google.com/recaptcha/api/siteverify';
>     $recaptcha_secret = 'YOUR_RECAPTCHA_SECRET_KEY';
>     $recaptcha_response = $_POST['recaptcha_response'];
> 
>     // Make and decode POST request:
>     $recaptcha = file_get_contents($recaptcha_url . '?secret=' . $recaptcha_secret . '&response=' . $recaptcha_response);
>     $recaptcha = json_decode($recaptcha);
> 
>     // Take action based on the score returned:
>     if ($recaptcha->score >= 0.5) {
>         // Verified - send email
>     } else {
>         // Not verified - show form error
>     }
> ```

Hi @Frenchomatic 

I see you just pasted an example of backend, what's the part you don't actually get, in v3 you choose what to do depending on the score value, if it meets your criteria, then, `// Verified - send email` accept the contact lead, else, treat it as spam ` // Not verified - show form error`.

**Comment by Frenchomatic** [comment]
Well a simple example of how it is used would be good. In other words here is a simple html form which is AMP compliant, here is the php form which sends back the response and forwards the email.

**Comment by ValentinBrclz** [comment]
@Frenchomatic : This is an issue, not a support group. I recognise that the documentation can be improved but spamming the issue with your *personal* problem is not going to solve that.

I disliked your comments because I got fed up of it and I'm not the only one. However, I would be happy to help if your personal problem was described *completely* somewhere where it is supposed to be (stackoverflow?) or if you open an issue to improve the documentation. You can DM me on Twitter if that is the case.

**Comment by Frenchomatic** [comment]
Grow up young man - you obviously don't understand what the definition of spamming means but  at least you recognize there is a deficiency. You have an arrogance issue and that is very much a personal problem which you should attempt to solve first - customers relationships is obviously not your strong point. 

Given GITHUB hosts the developers, they are in the best position to offer a blog post or whatever in the way of better supporting their product. I simply prompt them to do so as they have an important product which needs to be better explained so mere mortals can use it.

The whole point of github is the have software which people can share and **USE**.

**Comment by nainar** [comment]
@Frenchomatic your comments above are violating AMP's [Code of Conduct](https://github.com/ampproject/meta/blob/master/CODE_OF_CONDUCT.md) which aim to promote civil discourse on all AMP channels.

Please ensure that all future comments follow the Code of Conduct. Technical critique of features is appreciated - however, personally attacking any of our other members is not. 

Since this issue

**Comment by Frenchomatic** [comment]
Waste of time - just arrogance.

**Comment by nainar** [comment]
The above user has been blocked from posting on this project. 

While the initial comments have been deleted, we nonetheless apologize to the others in this thread.

**Comment by nickmoline** [comment]
@Frenchomatic 

I'm sorry you are frustrated, but this is not the reCAPTCHA project, documentation about the reCAPTCHA project is completely separate from the documentation on AMP.  What is documented on the AMP project is how to use the AMP Component, validation of reCAPTCHA is the same regardless if the post came from AMP or not, and unless someone from reCAPTCHA happens to be lurking on these issues, the people on the AMP team (and volunteers) are not equipped to answer questions outside of the scope of AMP.

The main documentation for reCAPTCHA V3 (the amp component uses V3) can be found here:
https://developers.google.com/recaptcha/docs/v3

And documentation about validating responses from recaptcha servers can be found here:
https://developers.google.com/recaptcha/docs/verify

The reCAPTCHA project has their own forum with their own set of community volunteers who try to help people available here:
https://groups.google.com/forum/#!forum/recaptcha

The code snippet you pasted in your previous post is well on its way towards validating the response, there's comments where your code would be, it takes the token included in the form post fields, sends it on to the reCAPTCHA validation service, and then has an if statement to determine what to do with the post depending on  that score.  In the sample code there is an ```if ($recaptcha->score >= 0.5) {``` and inside that is where you'd do your normal processing of the form, and in the else statement is what you'd do if they didn't validate the captcha submission.
